src/backend/distributed/planner/recursive_planning.c

```diff
@@ -1337,6 +1387,219 @@ NodeContainsSubqueryReferencingOuterQuery(Node *node)
 }
 
 
+/*
+ * ReplaceRTERelationWithRteSubquery replaces the input rte relation target entry
+ * with a subquery. The function also pushes down the filters to the subquery.
+ *
+ * It then recursively plans the subquery. This subquery is wrapped with another subquery
+ * as a trick to reduce network cost, because we currently don't have an easy way to
+ * skip generating NULL's for non-required columns, and if we create (SELECT a, NULL, NULL FROM table)
+ * then this will be sent over network and NULL's also occupy some space. Instead of this we generate:
+ * (SELECT t.a, NULL, NULL FROM (SELECT a FROM table) t). The inner subquery will be recursively planned
+ * but the outer part will not be yet it will still have the NULL columns so that the query is correct.
+ */
+void
+ReplaceRTERelationWithRteSubquery(RangeTblEntry *rangeTableEntry,
+								  List *requiredAttrNumbers,
+								  RecursivePlanningContext *context)
+{
+	List *outerQueryTargetList = NIL;
+	Query *subquery = WrapRteRelationIntoSubquery(rangeTableEntry, requiredAttrNumbers,
+												  &outerQueryTargetList);
+	List *restrictionList =
+		GetRestrictInfoListForRelation(rangeTableEntry,
+									   context->plannerRestrictionContext);
+	List *copyRestrictionList = copyObject(restrictionList);
+	Expr *andedBoundExpressions = make_ands_explicit(copyRestrictionList);
+	subquery->jointree->quals = (Node *) andedBoundExpressions;
+
+	/*
+	 * Originally the quals were pointing to the RTE and its varno
+	 * was pointing to its index in rtable. However now we converted the RTE
+	 * to a subquery and the quals should be pointing to that subquery, which
+	 * is the only RTE in its rtable, hence we update the varnos so that they
+	 * point to the subquery RTE.
+	 * Originally: rtable: [rte1, current_rte, rte3...]
+	 * Now: rtable: [rte1, subquery[current_rte], rte3...] --subquery[current_rte] refers to its rtable.
+	 */
+	UpdateVarNosInNode(subquery, SINGLE_RTE_INDEX);
```

Comment:
the naming of this function suggests it takes a Node, but it actually takes a Query and then only looks at a specific part of the query.

Reply:
I think we can just pass the quals here